Make binlog support non-blocking connection

andreas-bok-sociomantic opened the pull request:
The PR fixes the issue #34 where binlog can't be used with a NON_BLOCKING connection.

ben-palmer-sociomantic commented on tests/unit/binlog_async.cc:
```diff
@@ -0,0 +1,118 @@
+  while (ret != DRIZZLE_RETURN_OK)
+  {
+    drizzle_wait(con);
+    ret = drizzle_connect(con);
```
Could this while loop just keep spinning?

ben-palmer-sociomantic commented on tests/unit/binlog_async.cc:
```diff
@@ -0,0 +1,118 @@
+
+  binlog= drizzle_binlog_init(con, binlog_event, binlog_error, (void*)&cxt, true);
+  ret= drizzle_binlog_start(binlog, 0, "", 0);
+  SKIP_IF_(ret == DRIZZLE_RETURN_ERROR_CODE, "Binlog is not open?: %s(%s)",
```
Does this mean that if there is a problem with the binlog opening it just skips this test? Is that the intended behaviour?

ben-palmer-sociomantic commented:
Should this be against the `v5.3.x` branch?

andreas-bok-sociomantic commented:
> Should this be against the v5.3.x branch?

I am sure. In a sense it's a bugfix which extends the use case for binlog but not the api itself. Don't know how that translates to versioning guidelines

ben-palmer-sociomantic commented:
> I am sure. In a sense it's a bugfix which extends the use case for binlog but not the api itself. Don't know how that translates to versioning guidelines

Ah good point, in that case it should stay like it is.

andreas-bok-sociomantic commented:
updated based on comments from @ben-palmer-sociomantic 

- code formatting 
- remove obsolete code 
- fix risk of infinite loop

```diff
diff --git a/tests/unit/binlog_async.cc b/tests/unit/binlog_async.cc
index b018bd9..080bc9d 100644
--- a/tests/unit/binlog_async.cc
+++ b/tests/unit/binlog_async.cc
@@ -45,7 +45,8 @@
 void binlog_error(drizzle_return_t ret, drizzle_st *connection, void *context)
 {
   (void) context;
-  ASSERT_EQ_(DRIZZLE_RETURN_EOF, ret, "%s(%s)", drizzle_error(connection), drizzle_strerror(ret));
+  ASSERT_EQ_(DRIZZLE_RETURN_EOF, ret, "%s(%s)", drizzle_error(connection),
+    drizzle_strerror(ret));
 }
 
 void binlog_event(drizzle_binlog_event_st *event, void *context)
@@ -90,16 +91,20 @@ int main(int argc, char *argv[])
                       getenv("MYSQL_SCHEMA"), opts);
   ASSERT_NOT_NULL_(con, "Drizzle connection object creation error");
 
-  drizzle_set_verbose(con, DRIZZLE_VERBOSE_INFO);
-
   ret= drizzle_connect(con);
 
   while (ret != DRIZZLE_RETURN_OK)
   {
-    drizzle_wait(con);
+    ret = drizzle_wait(con);
+    // Assert that connection is not broken while waiting for IO
+    ASSERT_EQ_(DRIZZLE_RETURN_OK, ret, "%s(%s)", drizzle_error(con), drizzle_strerror(ret));
+
+    // This will eventually return true
     ret = drizzle_connect(con);
   }
 
+  ASSERT_EQ_(DRIZZLE_RETURN_OK, ret, "%s(%s)", drizzle_error(con), drizzle_strerror(ret));
+
   int cxt = 1;
 
   binlog= drizzle_binlog_init(con, binlog_event, binlog_error, (void*)&cxt, true);
@@ -107,10 +112,7 @@ int main(int argc, char *argv[])
   SKIP_IF_(ret == DRIZZLE_RETURN_ERROR_CODE, "Binlog is not open?: %s(%s)",
     drizzle_error(con), drizzle_strerror(ret));
 
-  while (ret == DRIZZLE_RETURN_IO_WAIT)
-  {
-    ret = drizzle_wait(con);
-  }
+  ret= drizzle_wait(con);
 
   ASSERT_EQ_(DRIZZLE_RETURN_OK, ret, "Drizzle binlog start failure: %s(%s)",
     drizzle_error(con), drizzle_strerror(ret));
```

ben-palmer-sociomantic commented:
LGTM, @nemanja-boric-sociomantic do you have any feedback on this one?

andreas-bok-sociomantic commented:
Just added blocked. Would like to do a bit more testing with dive. Still open for comments though.

nemanja-boric-sociomantic commented:
Ah congratulations!

I'll review it today!

nemanja-boric-sociomantic commented:
I haven't looked into tests, but the fix itself LGTM!

ben-palmer-sociomantic commented:
> Just added blocked. Would like to do a bit more testing. Still open for comments though.

Hows the testing going? We good to unblock this now?

andreas-bok-sociomantic commented:
> Hows the testing going? We good to unblock this now? 

Nope my fix did not work correctly with the client `epoll` 

So this might go into the next v5.2.x point release. Currently there are already 12 closed PRs for the v5.2.x milestone.

andreas-bok-sociomantic committed:
Make binlog support non-blocking connection

The fix changes the behavior of the api function
drizzle_binlog_start such that, if the
connection is non-blocking, the client
is expected to read data directly from the file
descriptor obtained from drizzle_fd()

For blocking connections the behavior is the same

Addresses #34

andreas-bok-sociomantic commented:
ping ready for review

ben-palmer-sociomantic commented:
The commit message has a `wuch` in it and also a double `the`. Otherwise this LGTM.

andreas-bok-sociomantic commented:
Fixed typos in ci msg.

nemanja-boric-sociomantic commented:
LGTM

ben-palmer-sociomantic commented:
This pull request has been rebased via `git hub pull rebase`. Original pull request HEAD was b47949c3568199fd6d71ec8c9e512ae149fac2cb, new (rebased) HEAD is 7baf1f0260aaa02ed96a2d3262bfa26b5c48431d
